File: BFS/Basics/279. Perfect Squares/279. Perfect Squares.java
class Solution {
    public int numSquares(int n) {
        
        boolean[] visited = new boolean[n + 1];
        List<Integer> squareNumbers = new ArrayList<>();
        for (int i = 1; i <= Math.sqrt(n); i++) {
            if (i % 1 == 0) {
                squareNumbers.add(i * i);
                // System.out.println(i*i);
            }
        }
        
        Queue<Integer> queue = new LinkedList<>();
        queue.add(n);
        
        int res = -1;
        while (!queue.isEmpty()) {
            
            res++;
            int len = queue.size();
            while (len-- > 0) {
                
                int num = queue.poll();
                
                if (visited[num]) continue;
                visited[num] = true;
                
                if (num == 0) {
                    return res;
                }
                
                for (int squareNumber : squareNumbers) {
                    int diff = num - squareNumber;
                    if (diff < 0) continue;
                    if (visited[diff]) continue;
                    queue.add(diff);
                }
            }
        }
        return -1;
    }
}

    

File: BFS/Basics/279. Perfect Squares/279. Perfect Squares.cpp
class Solution {
public:
    int numSquares(int n) {
        
         int visited[n + 1];
        for (int i = 0; i <= n; i++) {
            visited[i] = 0;
        }
        vector<int> squareNumbers;
        for (int i = 1; i <= sqrt(n); i++) {
            if (i % 1 == 0) {
                squareNumbers.push_back(i*i);
            }
        }
        queue<int> q;
        q.push(n);
        
        int res = -1;
        while (!q.empty()) {
            
            res++;
            int len = q.size();
            while (len--) {
                
                int num = q.front();
                q.pop();
                
                if (visited[num] == 1) continue;
                visited[num] = 1;
                // cout << num << endl;
                
                if (num == 0) {
                    return res;
                }
                
                for (const auto& squareNumber : squareNumbers) {
                    
                    int diff = num - squareNumber;
                    if (diff < 0) continue;
                    if (visited[diff]) continue;
                    q.push(diff);
                }
            }
        }
        return -1;
    }
};